src/dsp_tools/commands/excel2json/new_lists.py

```diff
@@ -12,33 +12,105 @@
+            for i, row in df.iterrows():
+                if row["id"] in id_set:
+                    df.at[i, "id"] = _construct_non_duplicate_id_string(df.iloc[int(str(i))], preferred_lang)
+    return excel_dfs
+
+
+def _fill_one_df_auto_id_resolve_duplicates(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
     if "ID (optional)" not in df.columns:
         df["ID (optional)"] = pd.NA
-    df = _fill_auto_id_column(df, preferred_language)
+    df = _fill_one_df_auto_id(df, preferred_language)
     df["id"] = df["ID (optional)"].fillna(df["auto_id"])
-    df = _fill_parent_id(df, preferred_language)
+    df = _resolve_duplicate_custom_and_auto_id(df, preferred_language)
+    return df
+
+
+def _resolve_duplicate_custom_and_auto_id(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+    if (duplicate_filter := df["id"].duplicated(keep=False)).any():
+        for i in duplicate_filter.index[duplicate_filter]:
+            if pd.isna(df.at[i, "ID (optional)"]):
+                df.loc[i, "id"] = _construct_non_duplicate_id_string(df.iloc[i], preferred_language)
     return df
 
 
-def _fill_auto_id_column(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+def _fill_one_df_auto_id(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
     """For every node without manual ID, take the label of the preferred language as ID."""
     df["auto_id"] = pd.NA
     if not df["ID (optional)"].isna().any():
         return df
     if pd.isna(df.at[0, "ID (optional)"]):
-        df.at[0, "auto_id"] = df.at[0, f"{preferred_language}_list"]
+        df.loc[0, "auto_id"] = df.at[0, f"{preferred_language}_list"]
     columns = sorted(_get_columns_of_preferred_lang(df.columns, preferred_language, r"\d+"), reverse=True)
     for i, row in df.iterrows():
         if pd.isna(row["ID (optional)"]):
             for col in columns:
                 if pd.notna(row[col]):
                     df.at[i, "auto_id"] = row[col]
                     break
+    df = _resolve_one_df_auto_duplicate_ids(df, preferred_language)
     return df
 
 
-def _fill_parent_id(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+def _resolve_one_df_auto_duplicate_ids(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+    if (duplicate_filter := df["auto_id"].dropna().duplicated(keep=False)).any():
+        for i in duplicate_filter.index[duplicate_filter]:
+            df.at[i, "auto_id"] = _construct_non_duplicate_id_string(df.iloc[i], preferred_language)
+    return df
+
+
+def _construct_non_duplicate_id_string(row: pd.Series[Any], preferred_language: str) -> str:
+    columns = _get_columns_of_preferred_lang(row.index, preferred_language, r"\d+")
+    columns.insert(0, f"{preferred_language}_list")
+    id_cols = [row[col] for col in columns if pd.notna(row[col])]
+    return ":".join(id_cols)
+
+
+def _fill_one_df_parent_id(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
```

Comment:
renaming suggestion: _create_col_with_parent_id, or _create_col_with_parent_id_in_one_df. But I think "in one df" is obsolete, because only 1 df is passed as argument